### PR TITLE
Cache index queries

### PR DESCRIPTION
**Pull request by gouthamve**

### CODE REVIEW

File: pkg/chunk/storage/caching_storage_client.go [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+
+	return &cachingStorageClient{
+		StorageClient: client,
+		cache:         chunk.NewFifoCache(size),
+		validity:      validity,
+	}
+}
+
+func (s *cachingStorageClient) QueryPages(ctx context.Context, query chunk.IndexQuery, callback func(result chunk.ReadBatch) (shouldContinue bool)) error {
+	if len(query.RangeValuePrefix) != 0 || len(query.RangeValueStart) != 0 { // Only caching queries that read full rows.
```
Comment (gouthamve) [inline pkg/chunk/storage/caching_storage_client.go]: @tomwilkie Why do we need to do this?

File: pkg/chunk/storage/caching_storage_client.go [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+		return s.StorageClient.QueryPages(ctx, query, callback)
+	}
```
Comment (tomwilkie) [inline pkg/chunk/storage/caching_storage_client.go]: What % of queries does this skip?  Looks like we might need to cache them, as we do a lot of them: https://github.com/grafana/cortex/blob/master/pkg/chunk/schema.go#L709

File: pkg/chunk/storage/caching_storage_client.go [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+
+	value, updated, ok := s.cache.Get(query.TableName + query.HashValue)
```
Comment (tomwilkie) [inline pkg/chunk/storage/caching_storage_client.go]: Can we guarantee there will be no accidental collisions by concatenating these keys?  Should we include a separator?

File: pkg/chunk/storage/caching_storage_client.go [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+	if ok && time.Now().Sub(updated) < s.validity {
```
Comment (tomwilkie) [inline pkg/chunk/storage/caching_storage_client.go]: We have similar logic in other places where we use the FIFOCache, might be better to move it there.

File: pkg/chunk/fifo_cache.go [hunk 5]
```diff
@@ -5,9 +5,9 @@ import (
 	"time"
 )
 
-// fifoCache is a simple string -> interface{} cache which uses a fifo slide to
+// FifoCache is a simple string -> interface{} cache which uses a fifo slide to
 // manage evictions.  O(1) inserts and updates, O(1) gets.
-type fifoCache struct {
+type FifoCache struct {
```
Comment (tomwilkie) [inline pkg/chunk/fifo_cache.go]: Would be good to instrument this cache directly for hits/missed/evictions.

**Comment by gouthamve** [comment]
@tomwilkie Addressed all your feedback. Could you check the cache metrics. I might have been a little overzealous when I instrumented :)

File: pkg/chunk/fifo_cache.go [hunk 6]
```diff
@@ -3,18 +3,77 @@ package chunk
...
+	cacheStaleGets = promauto.NewCounterVec(prometheus.CounterOpts{
+		Namespace: "querier",
+		Subsystem: "cache",
+		Name:      "stale_gets_total",
+		Help:      "The total number of Get calls that had an entry which expired",
+	}, []string{"cache"})
 )
 
-// fifoCache is a simple string -> interface{} cache which uses a fifo slide to
+type cacheMetrics struct {
```
Comment (tomwilkie) [inline pkg/chunk/fifo_cache.go]: You should just put these in the struct; the extra type isn't of much use.

File: pkg/chunk/fifo_cache.go [hunk 7]
```diff
@@ -106,8 +183,17 @@ func (c *fifoCache) get(key string) (value interface{}, updated time.Time, ok bo
...
+		if time.Now().Sub(updated) < c.validity {
+			value = c.entries[index].value
+			return
+		}
+
+		c.metrics.totalMisses.Inc()
+		c.metrics.staleGets.Inc()
+		ok = false
+	} else {
+		c.metrics.totalMisses.Inc()
```
Comment (tomwilkie) [inline pkg/chunk/fifo_cache.go]: Try to un-nest this a little more:

```go
index, ok = c.index[key]
if !ok {
  c.metrics.totalMisses.Inc()
  return
}

updated := c.entries[index].updated
if time.Now().Sub(updated) > c.validity {
  c.metrics.totalMisses.Inc()
  c.metrics.staleGets.Inc()
  ok = false
  return
}

value = c.entries[index].value
return
}
```

Also might be worth getting rid of the named return values at this point.

File: pkg/chunk/storage/caching_storage_client.go [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+	}
+}
+
+func queryKey(q chunk.IndexQuery) string {
+	const sep = "\xff"
+	return q.TableName + sep +
+		q.HashValue + sep +
+		string(q.RangeValuePrefix) + sep +
+		string(q.RangeValueStart) + sep +
+		string(q.ValueEqual)
```
Comment (tomwilkie) [inline pkg/chunk/storage/caching_storage_client.go]: The RangeValueStart is often based on the query start time, so will be pretty much uncachable.  I think you should always query + cache the entire row, and filter cache hits on return.

File: pkg/chunk/storage/factory.go [hunk 9]
```diff
@@ -28,26 +32,32 @@ func (cfg *Config) RegisterFlags(f *flag.FlagSet) {
...
 	cfg.GCPStorageConfig.RegisterFlags(f)
 	cfg.CassandraStorageConfig.RegisterFlags(f)
+
+	f.IntVar(&cfg.IndexCacheSize, "store.index-cache-size", 0, "Size of in-memory index cache, 0 to disable.")
+	f.DurationVar(&cfg.IndexCacheValidity, "store.index-cache-validity", 15*time.Minute, "Period for which entries in the index cache are valid.")
 }
 
 // NewStorageClient makes a storage client based on the configuration.
-func NewStorageClient(cfg Config, schemaCfg chunk.SchemaConfig) (chunk.StorageClient, error) {
+func NewStorageClient(cfg Config, schemaCfg chunk.SchemaConfig) (client chunk.StorageClient, err error) {
```
Comment (tomwilkie) [inline pkg/chunk/storage/factory.go:41]: might be easier to introduce a function that calls `NewStoreClient`, like `NewCachingStorageClient`...

**Comment by gouthamve** [comment]
Incorporated most of the feedback, need to add tests.

**Comment by gouthamve** [comment]
The tests currently don't test hitting the same store again with the query. Will add that tomorrow. But this mostly confirms that everything is working.

File: pkg/chunk/storage/factory.go [hunk 10]
```diff
@@ -34,7 +34,7 @@ func (cfg *Config) RegisterFlags(f *flag.FlagSet) {
 	cfg.CassandraStorageConfig.RegisterFlags(f)
 
 	f.IntVar(&cfg.IndexCacheSize, "store.index-cache-size", 0, "Size of in-memory index cache, 0 to disable.")
-	f.DurationVar(&cfg.IndexCacheValidity, "store.index-cache-validity", 15*time.Minute, "Period for which entries in the index cache are valid.")
+	f.DurationVar(&cfg.IndexCacheValidity, "store.index-cache-validity", 15*time.Minute, "Period for which entries in the index cache are valid. Should be no higher than -ingester.max-chunk-idle")
```
Comment (tomwilkie) [inline pkg/chunk/storage/factory.go]: Full stop at end pls!

Comment (gouthamve) [reply pkg/chunk/storage/factory.go]: Done.

**Pull request merged**